client/src/interface/components/ChangePermission/index.js

```diff
@@ -1,192 +1,203 @@
+            name: null,
+            mode: null,
+            modeCheck: null,
+            modeInput: null,
+            applyRecursively: false
         };
     }
 
     handleSubmit() {
-        axios.patch(`/sftp_remode/${this.state.session_id}`, {
-            cwd: this.state.cwd + "/" + this.state.id, 
-            new_mode: this.state.mode
-        }).then(response => {
-            console.log(response.data);
-            window.location.reload();
-        }).catch(err => {
-            console.log('error', err);
-        });
+        sftp_chmod(
+            this.props.fm,
+            this.cwd,
+            this.state.name,
+            this.state.mode & 0o777,
+            this.state.applyRecursively)
     }
 
     handleClose = (ev) => {
-        if (ev.target && ev.target.id === 'button_save') {
+        if (ev.target && ev.target.id === 'button_apply') {
             this.handleSubmit();
         }
-        this.props.onChangePermissionClose();
+
+        this.props.fm.setState({
+            changePermissionOpen: false
+        })
     };
-    
-    handleOnChange = (mode_bit) => {
-        this.setState({mode: this.state.mode ^ mode_bit})
+
+    handleModeChange = (ev) => {
+        const bit = parseInt(ev.target.name)
+        this.setState({
+            mode: this.state.mode ^ bit,
+            modeCheck: this.state.modeCheck ^ bit,
+            modeInput: (this.state.modeCheck ^ bit).toString(8).padStart(3, '0')
+        })
     }
 
-    update_id_mode = (_cwd, _id, _mode, _session_id) => {
-        this.setState({cwd: _cwd,
-                       id: _id, 
-                       mode: _mode,
-                       session_id: _session_id})
+    handleModeInputChange = (ev) => {
+        const value = ev.target.value
+        if (value.length > 3) {
+            return
+        }
+        if (value.length !== 0 && !/^[0-7]+$/.test(value)) {
+            return
+        }
+        this.setState({
+            modeInput: value,
+            modeCheck: parseInt(value.padEnd(3, '0'), 8)
+        })
+    }
+    handleModeInputBlur = (_) => {
+        if (this.state.modeInput.length !== 3) {
+            this.setState({
+                modeCheck: (this.state.mode & 0o777),
+                modeInput: (this.state.mode & 0o777).toString(8).padStart(3, '0')
+            })
+        } else {
+            this.setState({
+                mode: this.state.mode & 0xfe00 | this.state.modeCheck
+            })
+        }
+    }
+    update_id_mode = (_cwd, _name, _mode) => {
+        this.cwd = _cwd
+        this.setState({
+            name: _name,
+            mode: _mode,
+            modeCheck: (_mode & 0o777),
+            modeInput: (_mode & 0o777).toString(8).padStart(3, '0')
+        })
     }
 
     render() {
         const {open} = this.props;
+        if (!open) {
+            return null
+        }
+
+        const {name, mode, modeCheck, modeInput, applyRecursively} = this.state
+        const entryIsDir = isDir(mode)
+
```

Comment:
instead of writing out all of the 9 checkboxes, I refactored them into a list of the elements so that we can change the styles of the checkboxes in one place.
let me know if you dislike this approach and we can discuss that